BMAG v2.0.0: Rewrite

uptudev opened the pull request:
Completely rewrote the mod from scratch, adding a binding menu which allows binding any key or button to any function. ~~Default settings are still the same~~, but you can now rebind any of the mod functions.

**All** old GPLv3-licensed code was replaced with new MIT/X11-licensed code, as GPLv3 prohibits shipping software with proprietary code, which Balatro might contain. Steamodded metadata has been updated to the new v1.0.0 format, and a `CHANGELOG.md`, `CODE_OF_CONDUCT.md` and `CONTRIBUTING.md` have been added to track changes and outline contribution guidelines.

Files in `./.dev` are used to bootstrap the development process, and shouldn't be included in release archives. The only files that should be included in the releases are:

* `./localization/`,
* `./src/`,
* `./bmag.json`,
* `./LICENSE`, and
* `./README.md`.

To preview, I have v2.0.0 as a release on my fork of the repo [here](https://github.com/uptudev/bmag/releases/tag/v2.0.0). I will be testing it out in the meantime to ensure no bugs slipped through, but all code is documented and straightforward.

uptudev committed:
Refactor pt.1
Taking a break for a bit. There will be a part 2 for this commit.

Mostly consists of restructuring the program as well as changing
variable names to be more accurate. Might revert some in pt.2

uptudev committed:
Refactor part 2 Electric Boogaloo
Slowly but surely this codebase will make sense to me

uptudev committed:
Refactor part 3

uptudev committed:
Merge pull request #1 from uptudev/refactor

Refactor codebase

uptudev committed:
Replace old links in README to forked version

uptudev committed:
Clarify installation instructions

uptudev committed:
Update README.md

Remove Handy mention

uptudev committed:
Update en-us.lua

`sed` messed up one of the localization strings

uptudev committed:
`sed` fucked me again

uptudev committed:
Merge branch 'main' of github.com:uptudev/bmag

uptudev committed:
Implement queues properly

uptudev committed:
Update issue templates

Create issue templates for GitHub tracking

uptudev committed:
Merge pull request #2 from uptudev/add-github-issue-templates

Update issue templates

uptudev committed:
Merge pull request #6 from Kooluve/main

Merge pull request #7 from uptudev/main

uptudev committed:
Remove notes and legacy code
- Notes on v2 development can be found in the GitHub Issues.
- Legacy code removed.

uptudev committed:
Update issue templates

Added headers to make reading issues easier

uptudev committed:
Merge pull request #8 from uptudev/add-gh-issue-headers

Update issue templates

uptudev committed:
chore(legal): delete all GPL-restricted code and add MIT/X11 license

modify:
* `./LICENSE`: replace GPLv3 license with modified MIT/X11 license.

delete:
- `./README_JP.md`: remove old docs not written by me
- `./README_ZH.md`: remove old docs not written by me
- `./config.lua`: remove old GPLv3 licensed code
- `./localization/en-us.lua`: remove old GPLv3 licensed code
- `./localization/ja.lua`: remove old GPLv3 licensed code
- `./localization/zh_CN.lua`: remove old GPLv3 licensed code

Signed-off-by: uptu <uptu@uptu.dev>

uptudev committed:
refactor(main): move main entrypoint to `./src/main.lua`

create:
+ `./src/main.lua`: add main source code file

modify:
* `./bmag.lua`: change to only contain Steamodded metadata and a
  `load_file` call to `./src/main.lua`

Signed-off-by: uptu <uptu@uptu.dev>

uptudev committed:
feat(docs): add CHANGELOG

create:
+ `./CHANGELOG.md`: add a CHANGELOG compliant with the standards set out
  at https://keepachangelog.com/

Signed-off-by: uptu <uptu@uptu.dev>

uptudev committed:
refactor(libs): add Queue and Timer objects to `./src/libs/`

create:
+ `./src/libs/timers.lua`: add an implementation for a table for
  tracking monotonic keypress timers

modify:
* `./queue.lua` -> `./src/libs/queue.lua`: move the queue implementation
  to the `libs` directory

Signed-off-by: uptu <uptu@uptu.dev>

uptudev committed:
chore(git): add git hooks to VCS

create:
+ `./.dev/hooks/commit-msg`: add hook to sign off on git commit messages
+ `./.dev/hooks/post-checkout`: add hook to run a bootstrapping script
  on checkout

uptudev committed:
chore(git): add default commit template

create:
+ `./.dev/commit-template`: add a default git commit template

uptudev committed:
fix(git): fix mistake in git hook

modify:
* `./.dev/hooks/commit-msg`: sign the commit message in-place

uptudev committed:
chore(git): add bootstrapping script

create:
+ `./.dev/bootstrap.sh`: add bootstrapping script

Signed-off-by: uptu <uptu@uptu.dev>

uptudev committed:
chore(git): patch bootstrapping script

modify:
* `./.dev/bootstrap.sh`: Error catching and pipeline fix

Signed-off-by: uptu <uptu@uptu.dev>

uptudev committed:
docs: add contribution guidelines

create:
+ `./CONTRIBUTING.md`: add contribution guidelines

Signed-off-by: uptu <uptu@uptu.dev>

uptudev committed:
chore(git): fix bootstrapping script again

modify:
* `./.dev/bootstrap.sh`: should be all fixed now ;-;

Signed-off-by: uptu <uptu@uptu.dev>

uptudev committed:
feat(git): add COC violations to GH issues templates

uptudev committed:
docs: add Code of Conduct

create:
+ `./CODE_OF_CONDUCT.md`: add the Code of Conduct to the repo, based off
  of Contributor Covenant v2.1

Signed-off-by: uptu <uptu@uptu.dev>

uptudev committed:
docs: fill in `CONTRIBUTING.md` template

modify:
* `./CONTRIBUTING.md`: fill in template

Signed-off-by: uptu <uptu@uptu.dev>

uptudev committed:
docs: make `CONTRIBUTING.md` compliant with RedHat style guide

modify:
* `./CONTRIBUTING.md`: make compliant with RedHat supplementary style
  guide (now linked in file)

Signed-off-by: uptu <uptu@uptu.dev>

uptudev committed:
docs: add script execution warning to `CONTRIBUTING.md`

modify:
* `./CONTRIBUTING.md`: add warning that running unfamiliar scripts is
  bad, m'kay?

Signed-off-by: uptu <uptu@uptu.dev>

uptudev committed:
docs: change list to ordered in `CONTRIBUTING.md`

modify:
* `./CONTRIBUTING.md`: change "Environment Setup" section to have an
  ordered list instead of an unordered list

Signed-off-by: uptu <uptu@uptu.dev>

uptudev committed:
docs: add IRC channel to potential CoC violation report channels

modify:
* `./CODE_OF_CONDUCT.md`: add IRC channel to Code of Conduct

Signed-off-by: uptu <uptu@uptu.dev>

uptudev committed:
feat(input): add InputMap class for holding keybinds

create:
+ `./src/libs/input_map.lua`: create InputMap class for storing and
  querying keybind information.

Signed-off-by: uptu <uptu@uptu.dev>

uptudev committed:
fix(metadata): replace legacy metadata definition with new standard

create:
+ `./bmag.json`: add new Steamodded JSON metadata

modify:
* `./CHANGELOG.md`: replace `./bmag.lua` with `./bmag.json`
* `./CONTRIBUTING.md`: replace `./bmag.lua` with `./bmag.json`

delete:
- `./bmag.lua`: remove the old metadata Lua file

Signed-off-by: uptu <uptu@uptu.dev>

uptudev committed:
refactor(input): move `input_map.lua` to `input.lua`

modify:
* `./src/libs/input.lua`: change filename from `./src/libs/input_map.lua`
* `./src/main.lua`: change references to `input_map.lua` to `input.lua`

Signed-off-by: uptu <uptu@uptu.dev>

uptudev committed:
feat(localization): scaffold localization

create:
+ `./localization/`: add the localization file directory
+ `./localization/en-us.lua`: create the English localization file

Signed-off-by: uptu <uptu@uptu.dev>

uptudev committed:
refactor(input): change `InputMap:get_feature()` to `InputMap:get()`

modify:
* `./src/libs/input.lua`: change the method name of
`InputMap:get_feature()` to `InputMap:get()`

Signed-off-by: uptu <uptu@uptu.dev>

uptudev committed:
feat(input): implement Love2D event hooks

create:
+ `./src/libs/l2d.lua`: add Love2D event hooks

modify:
* `./src/main.lua`: import `./src/libs/l2d.lua`

Signed-off-by: uptu <uptu@uptu.dev>

uptudev committed:
docs(coc): remove top level header

modify:
* `./CODE_OF_CONDUCT.md`: remove the first level header

Signed-off-by: uptu <uptu@uptu.dev>

uptudev committed:
docs(coc): make CoC compliant with RedHat SSG

modify:
* `./CODE_OF_CONDUCT.md`: make (mostly) compliant with RedHat SSG

Signed-off-by: uptu <uptu@uptu.dev>

uptudev committed:
docs(git): add GitHub PR template

create:
+ `./.github/pull_request_template.md`: add a pull request template

Signed-off-by: uptu <uptu@uptu.dev>

uptudev committed:
fix(metadata): fix broken `version` and import paths

modify:
* `./bmag.json`: fix broken `version` value
* `./src/main.lua`: fix broken library import paths

Signed-off-by: uptu <uptu@uptu.dev>

uptudev committed:
refactor(input): change names of `input.lua` and `l2d.lua`

modify:
* `./src/libs/bindmap.lua`: rename `InputMap` to `BindMap`
* `./src/main.lua`: rename `InputMap` to `BindMap` and change import
  paths for moved libraries

moved:
* `./src/libs/input.lua` -> `./src/libs/bindmap.lua`
* `./src/libs/l2d.lua` -> `./src/libs/input.lua`

Signed-off-by: uptu <uptu@uptu.dev>

uptudev committed:
fix(input): rename `input_map` to `bind_map`

modify:
* `./src/main.lua`: renamed `STATE.input_map` to `STATE.bind_map`
  (forgot after last commit)

Signed-off-by: uptu <uptu@uptu.dev>

uptudev committed:
feat(input): add `Binding` object for storing press/release functions

create:
+ `./src/libs/binding.lua`: add `Binding` object definition

modify:
* `./src/libs/bindmap.lua`: rename `feature` to `binding`
* `./src/libs/input.lua`: change `on_press`/`on_release` to
  `press`/`release`

Signed-off-by: uptu <uptu@uptu.dev>

uptudev committed:
docs(lua): replace all code documentation with `ldoc` compliant format

modify:
* `./src/libs/binding.lua`: replace all code documentation with `ldoc` compliant format
* `./src/libs/bindmap.lua`: replace all code documentation with `ldoc` compliant format
* `./src/libs/input.lua`: replace all code documentation with `ldoc` compliant format
* `./src/libs/queue.lua`: replace all code documentation with `ldoc` compliant format
* `./src/libs/timers.lua`: replace all code documentation with `ldoc` compliant format
* `./src/main.lua`: replace all code documentation with `ldoc` compliant format

Signed-off-by: uptu <uptu@uptu.dev>

uptudev committed:
fix(input): changed input system to go straight to timers from Love2D

modify:
* `./src/libs/binding.lua`: change `press/release` binding scheme to
  `click/hold`
* `./src/libs/input.lua`: change Love2D hooks to send `press/release`
events to the `TimerTable`, not the raw `Binding`s.

Signed-off-by: uptu <uptu@uptu.dev>

uptudev committed:
fix(input): change `touch` to `istouch`

modify:
* `./src/libs/input.lua`: change `touch` to `istouch`

Signed-off-by: uptu <uptu@uptu.dev>

uptudev committed:
feat(input): add `src/libs/bindings.lua` to imports

modify:
* `./src/main.lua`: change the contents of the file

Signed-off-by: uptu <uptu@uptu.dev>

uptudev committed:
chore(git): remove gitignore

delete:
- `./.gitignore`: remove gitignore; untracked files removed locally

Signed-off-by: uptu <uptu@uptu.dev>

uptudev committed:
docs(contrib): add project flowchart

modify:
* `./CONTRIBUTING.md`: add project flowchart

Signed-off-by: uptu <uptu@uptu.dev>

uptudev committed:
docs(contrib): add configuration menu to Mermaid graph

modify:
* `./CONTRIBUTING.md`: add configuration menu to Mermaid graph

Signed-off-by: uptu <uptu@uptu.dev>

uptudev committed:
refactor(input): remove arguments in `Binding` constructor

modify:
* `./src/libs/binding.lua`: remove arguments from `Binding:new()`

Signed-off-by: uptu <uptu@uptu.dev>

uptudev committed:
feat(input): add `is_bound` method to `BindMap`

modify:
* `./src/libs/bindmap.lua`: add `is_bound(button)` method to `BindMap`
  object.

Signed-off-by: uptu <uptu@uptu.dev>

uptudev committed:
refacator(input): add `hold_start` and `hold_end` functions

modify:
* `./src/libs/binding.lua`: add seperate `hold_start` and `hold_end`
  functions for handling some fun behaviour
* `./src/libs/input.lua`: add pause checks to Love2D overrides
* `./src/libs/timers.lua`: finish timer functionality

Signed-off-by: uptu <uptu@uptu.dev>

uptudev committed:
feat(input): add hook into update loop

modify:
* `./src/libs/bindmap.lua`: add click/hold threshold
* `./src/libs/timers.lua`: add hook into update loop

Signed-off-by: uptu <uptu@uptu.dev>

uptudev committed:
refactor(core): remove queues (no longer planned to be necessary)

modify:
* `./src/main.lua`: remove static queue and `./src/libs/queue.lua`
  imports

delete:
- `./src/libs/queue.lua`: remove queue implementation

Signed-off-by: uptu <uptu@uptu.dev>

uptudev committed:
feat(core): added most of the main features/endpoints

create:
+ `./.gitignore`: exclude Balatro exe dump from VCS
+ `./src/libs/feats.lua`: add main functionality of the mod (except
  multiselect because it's late rn)

Signed-off-by: uptu <uptu@uptu.dev>

uptudev committed:
feat(core): import feature functions in main file

modify:
* `./src/main.lua`: add `./src/libs/feats.lua` to imports

Signed-off-by: uptu <uptu@uptu.dev>

uptudev committed:
refactor(core): utilize delta time provided in update loop

modify:
* `./src/libs/bindmap.lua`: remove old timer threshold field
* `./src/libs/timers.lua`: change timers from monotonic frame counters
  to proper monotonic realtime clocks using delta time params

Signed-off-by: uptu <uptu@uptu.dev>

uptudev committed:
feat(input): add binding names/ids for proper indexing

modify:
* `./src/libs/binding.lua`: add names to the `Binding` class
* `./src/libs/bindmap.lua`: implement name indexing of `Binding`s

Signed-off-by: uptu <uptu@uptu.dev>

uptudev committed:
refactor(core): add config menu

create:
+ `./src/libs/config.lua`: add config menu

modify:
* `./localization/en-us.lua`: add config menu
* `./src/libs/binding.lua`: make compatible with config menu
* `./src/libs/bindmap.lua`: make compatible with config menu
* `./src/libs/input.lua`: make compatible with config menu
* `./src/libs/timers.lua`: make compatible with config menu
* `./src/main.lua`: make compatible with config menu

Signed-off-by: uptu <uptu@uptu.dev>

uptudev committed:
refactor(core): add input binding menu

modify:
* `./localization/en-us.lua`: add config menu localization keys
* `./src/libs/bindmap.lua`: refactor to work with new GUI config
* `./src/libs/config.lua`: implement GUI config (not persistent yet;
  TODO: make persistent)
* `./src/libs/feats.lua`: add multiselect function (and accidentally
  delete the doc header whoops)
* `./src/libs/input.lua`: refactor to properly handle multiselect
* `./src/libs/timers.lua`: refactor to properly handle multiselect
* `./src/main.lua`: add extra state for the multiselect functionality

delete:
- `./src/libs/binding.lua`: unnecessary class after refactor

Signed-off-by: uptu <uptu@uptu.dev>

uptudev committed:
fix(core): remove unnecessary state variable

modify:
* `./src/libs/feats.lua`: refactored to use `STATE.multiselect` over
  `STATE.first_multiselect_iter`
* `./src/libs/timers.lua`: refactored to use `STATE.multiselect` over
  `STATE.first_multiselect_iter`
* `./src/main.lua`: removed `STATE.first_multiselect_iter`

Signed-off-by: uptu <uptu@uptu.dev>

uptudev committed:
fix(feat): fix multiselect bug where first card would deselect itself

modify:
* `./src/libs/feats.lua`: multiselect bugfix

Signed-off-by: uptu <uptu@uptu.dev>

uptudev committed:
feat(input): add ability to overwrite already-bound keys

modify:
* `./src/libs/bindmap.lua`: add ability to overwrite already-bound keys

Signed-off-by: uptu <uptu@uptu.dev>

uptudev committed:
feat(input): make bindings persistent

modify:
* `./localization/en-us.lua`: add save button localization field
* `./src/libs/bindmap.lua`: add `from` generator to apply metatable
* `./src/libs/config.lua`: add save button
* `./src/main.lua`: add persistent map load on mod load

Signed-off-by: uptu <uptu@uptu.dev>

uptudev committed:
docs(localization): add `en-us` localization documentation

modify:
* `./localization/en-us.lua`: document localization file to make future
  localization easier for contributors

Signed-off-by: uptu <uptu@uptu.dev>

uptudev committed:
docs(core): update source documentation

modify:
* `./src/libs/bindmap.lua`: updated documentation
* `./src/libs/config.lua`: updated documentation
* `./src/libs/feats.lua`: updated documentation
* `./src/libs/timers.lua`: updated documentation
* `./src/main.lua`: updated documentation

Signed-off-by: uptu <uptu@uptu.dev>

uptudev committed:
docs(readme): update to v2 specs

modify:
* `./README.md`: update `README.md` to v2

Signed-off-by: uptu <uptu@uptu.dev>

uptudev committed:
fix(release): remove `alpha` tag from metadata SemVer

modify:
* `./bmag.json`: remove `alpha` prerelease tag

Signed-off-by: uptu <uptu@uptu.dev>

uptudev committed:
docs(changelog): update changelog for v2 release

modify:
* `./CHANGELOG.md`: add all major changes

Signed-off-by: uptu <uptu@uptu.dev>

uptudev committed:
fix(feats): fix play/discard allowing discards at 0 left

modify:
* `./src/libs/feats.lua`: fix play/discard not checking if there are
  hands or discards remaining

Signed-off-by: uptu <uptu@uptu.dev>

uptudev committed:
docs(changelog): update changelog to v2.0.1

modify:
* `./CHANGELOG.md`: add v2.0.1 bugfix

Signed-off-by: uptu <uptu@uptu.dev>

uptudev committed:
chore(metadata): update mod version to v2.0.1

modify:
* `./bmag.json`: update mod version to v2.0.1

Signed-off-by: uptu <uptu@uptu.dev>

uptudev committed:
chore(release): add release script

create:
+ `./.dev/release.sh`: add a release archive-generating script

modify:
* `./.gitignore`: add `./.releases` to ignore list

Signed-off-by: uptu <uptu@uptu.dev>

Kooluve commented:
Thank you for your hard work!
I apologize for the delay—I’ve been quite busy recently and may not have much time to review and merge PRs.
To be honest, since I rarely play Balatro these days, and given that Handy is overall a better mod, I wasn’t planning on updating this one anymore (though I probably should have mentioned that in the README). As long as the current functionality works, that’s fine for now. Honestly, I’d recommend using and maintaining Handy instead—it has more features and a stronger ecosystem, whereas my mod may not be worth optimizing further. That said, if you’d like to make further improvements, that would be amazing!
I’d love to hear your thoughts, though. Thanks again for your help!

uptudev committed:
fix(multiselect): fix minor bug causing accidental card click

modify:
* `./src/libs/feats.lua`: fix a bug where attempting to multiselect
  while holding over the background, then a card, then the background
  would cause the card to be unselected

Signed-off-by: uptu <uptu@uptu.dev>

uptudev committed:
docs(changelog): update changelog with new bugfixes

modify:
* `./CHANGELOG.md`: update changelog with new bugfixes

Signed-off-by: uptu <uptu@uptu.dev>

uptudev commented:
If you want to transfer the repo, I can continue as the active maintainer. This mod is listed in many threads that show up when you search Google for "balatro mods", so if it's deprecated as you say, I can continue development. Transferring through GitHub should keep all of the old links intact (redirecting them to the new location).

Handy is nice, but it is missing features which this mod does add. The version in this PR keeps all of the old functionality of BMAG while expanding it to allow binding any button, so I don't mind continuing on as the maintainer if you're done with it! The proposed v2 I've made is a rather large overhaul, but keeps all of the old functionality this mod has, which I feel that many do still enjoy (me included)!

Have a great day! :)

uptudev committed:
fix(play/discard): remove `print()` call left in while debugging

modify:
* `./src/libs/feats.lua`: remove 2 `print()` logging calls from
  debugging the quick play and quick discard system.

Signed-off-by: uptu <uptu@uptu.dev>

uptudev committed:
chore(docs): add video to `README.md`; reformat `CHANGELOG.md`

modify:
* `./CHANGELOG.md`: reformat to include some missing features
* `./README.md`: add a video preview of the mod at work

Signed-off-by: uptu <uptu@uptu.dev>

uptudev committed:
chore(docs): replace video link with GH embed

modify:
* `./README.md`: replace video link with GitHub built-in embedded video

Kooluve commented:
Thank you. I’ll consider merging or transferring this at a later time. That said, I still believe some of BMag’s unique features could be integrated into Handy without too much difficulty—this might be simpler than a full rewrite. I’d suggest reevaluating both approaches.
However, given the effort you’ve already put into these improvements, completing this project would still be worthwhile, even if it may not be strictly necessary.

Kooluve commented:
Apologies for the long delay in addressing this project.

I tested version 2.0.1 and noticed that a long right-click incorrectly triggers buttons outside the hand zone (e.g., "Sort Hand," "Play Hand," "View Deck," "Options," etc.). My original implementation included area restrictions to prevent this.

Additionally, I see you replaced the default Sort and Play/Discard key bindings. While the current version makes customization convenient, I’d like to note that most users are likely accustomed to the original key layout. Moreover, since Play/Discard is used more frequently than Sort, and many mice lack side buttons, assigning the scroll wheel to Play/Discard by default might be more practical.

uptudev committed:
fix(multiselect): fix multiselect working on more than just `Card`s

modify:
* `./src/libs/feat.lua`: add two missing `if` statements to ensure that
  the multiselect target is a `Card` and not a UI element like the Sort,
  Discard, or Play buttons.

Signed-off-by: uptu <uptu@uptu.dev>

uptudev committed:
fix(meta): increment version number to v2.0.2

modify:
* `./CHANGELOG.md`: add v2.0.2 changes
* `./bmag.json`: increment mod version to `2.0.2`

Signed-off-by: uptu <uptu@uptu.dev>

uptudev commented:
Patched that error with an `if` statement to ensure the target is a `Card`, as that should be easier computationally than checking the bounding box every frame that multiselect is processing.

Regarding the control switch, it's *extremely* easy to accidentally play or discard a hand by bumping the scroll wheel. The change was due to the fact that accidentally sorting your hand is much less detrimental to a run than accidentally playing or discarding a hand because you grabbed your mouse too quickly or bumped the wheel with your fingers.

Given the controls can be rebound much easier, I thought that having a default that is safe and less error prone may be best. If you want, I can change it later tonight, as it just involves swapping out the default table entries in `BindMap:new()`.

Kooluve commented:
You're absolutely right. I had considered the potential for accidental triggers before but didn't address it thoroughly. Let’s proceed with your suggested default setup.

I’ll continue with some additional testing and add localization support. If everything goes well, I plan to merge this pull request soon.

Kooluve commented:
A few additional minor issues:

1. right-click highlighting is too immediate
Currently, right-click triggers highlighting instantly, leaving no room for hesitation. Additionally, a simple right-click (without drag) causes the card to highlight and then immediately unhighlight, which feels redundant.
In my original implementation, highlighting only began after the cursor moved a certain distance after press down (e.g., Vector_Dist(R_cursor_down.T, C.cursor_hover.T) >= 0.1 * G.MIN_CLICK_DIST).

2. View Deck button is still triggered by right-click
The "View Deck" button in the bottom-right corner remains activatable via right-click, which should be restricted.

3. Right-click selects all card types unintentionally
Cards like Jokers, Tarot cards, and others (both in hand and in the shop) are currently affected by right-click selection, which may not be intended.

4. Game crashes when holding right-click and then left-clicking on any menu button

uptudev committed:
fix(feat): fix crash when attempting to multiselect over a UI element

modify:
* `./src/libs/feats.lua`: fix a crash when holding the `multiselect`
  bind down and clicking on a menu button

Signed-off-by: uptu <uptu@uptu.dev>

uptudev committed:
fix(feat): remove the ability to select non-hand cards with multiselect

modify:
* `./src/libs/feats.lua`: remove the ability to select non-hand cards
  with multiselect

Signed-off-by: uptu <uptu@uptu.dev>

uptudev committed:
fix(feat): only multiselect after user has moved mouse

modify:
* `./src/libs/feats.lua`: refactor multiselect to ensure multiselecting
  only happens after the user has moved the mouse
* `./src/libs/timers.lua`: set/unset multiselect start position on
  keypress
* `./src/main.lua`: add multiselect start position to mod state

Signed-off-by: uptu <uptu@uptu.dev>

uptudev commented:
>     1. right-click highlighting is too immediate
>        Currently, right-click triggers highlighting instantly, leaving no room for hesitation. Additionally, a simple right-click (without drag) causes the card to highlight and then immediately unhighlight, which feels redundant.
>        In my original implementation, highlighting only began after the cursor moved a certain distance after press down (e.g., Vector_Dist(R_cursor_down.T, C.cursor_hover.T) >= 0.1 * G.MIN_CLICK_DIST).

Fixed by [`f558f39`](https://github.com/Kooluve/Better-Mouse-And-Gamepad/pull/8/commits/f558f398ab9ffb69a1b26e90b76193af0f5e8ba6).
 
>     2. View Deck button is still triggered by right-click
>        The "View Deck" button in the bottom-right corner remains activatable via right-click, which should be restricted.
> 
>     3. Right-click selects all card types unintentionally
>        Cards like Jokers, Tarot cards, and others (both in hand and in the shop) are currently affected by right-click selection, which may not be intended.

Both fixed by [`3e85587`](https://github.com/Kooluve/Better-Mouse-And-Gamepad/pull/8/commits/3e855872cf5b657a9b9566ae72a2ab8284c22817).

>     4. Game crashes when holding right-click and then left-clicking on any menu button

Fixed by [`0a7434d`](https://github.com/Kooluve/Better-Mouse-And-Gamepad/pull/8/commits/0a7434d802a77194f65f050f05bd51d21c93a812).

___

Drafting [a new release](https://github.com/uptudev/bmag/releases/tag/v2.0.3) right now on my repo; let me know if you find any other issues :)

uptudev committed:
fix(meta): increment version number and update CHANGELOG

modify:
* `./CHANGELOG.md`: update with `v2.0.3` changes
* `./bmag.json`: increment version number

Signed-off-by: uptu <uptu@uptu.dev>

uptudev commented on src/libs/feats.lua:
```diff
@@ -7,22 +7,26 @@ function multiselect_hold()
         if
             G.CONTROLLER.hovering.target ~= nil and
-            G.CONTROLLER.hovering.target:is(Card)
+            G.CONTROLLER.hovering.target:is(Card) and not
```
This `and not` is erroneous and patched by [`f558f39`](https://github.com/Kooluve/Better-Mouse-And-Gamepad/pull/8/commits/f558f398ab9ffb69a1b26e90b76193af0f5e8ba6), which simplifies the logic. It is an artifact accidentally left in from moving lines around during experimentation.

Also appears on line 18, likewise patched in the next commit.

uptudev committed:
fix(feat): fix error caused by the game overwriting `prev_target`

modify:
* `./src/libs/feats.lua`: fix a minor error introduced during
  multiselect refactor where some state variables wouldn't be populated
  correctly (unlikely but a possible bug regardless)
* `./src/main.lua`: add `prev_target` to mod state variables to use
  instead of `G.CONTROLLER.hovering.prev_target`, which the engine
  overwrites during gameplay, causing erroneous behaviour where one
  could spam-click a card by putting the cursor over the blank space a
  card moves into/away from when selected

Signed-off-by: uptu <uptu@uptu.dev>

uptudev committed:
fix(meta): increment version number

modify:
* `./CHANGELOG.md`: add `v2.0.4` changes
* `./bmag.json`: increment version number to `v2.0.4`

Signed-off-by: uptu <uptu@uptu.dev>

uptudev commented:
Fixed a minor error caused by using game state variables, which allowed the user to spam a card if they got their cursor in the right spot. It's unlikely to cause issues in gameplay, but I fixed it anyways.

There was also one more issue that I found where I assign the `prev_prev_target` in the wrong branch of my `if` statement. I can't find any more bugs or errors after fixing it and reviewing my code, so let me know if any more come up and I'll patch them out.

(ignore the force-push; I forgot to update the changelog correctly, then forgot to add it to the commit lol)

Kooluve commented:
Thank you for your efforts, all the previous issues have been resolved. However, I've noticed a few problems with the keybinds:
1. The multiselect keybinding cannot be modified.
2. The "restart run" action cannot be triggered with any key.

Additional suggestions:
1. Add a "Restore Default Keybinds" button.
I've noticed that for many mods, even deleting the mods, mod loader, and config folder, mods's settings still aren't reset. A dedicated reset button would be helpful.
2. Limit bindable keys to common game keys (letters, numbers, spacebar, etc.).
Modifier keys (like Ctrl, Alt) or system keys (Win, CapsLock) can cause unintended issues such as StickyKeys activation or unexpected behavior.
3. Rapid mouse movement during multiselect causes intermittent card selection.
This seems to occur because the game updates input on a per-frame basis, and fast movement may be missed. This issue exists in both my version and Handy.
Since most users are unlikely to encounter this (or can avoid it by moving more slowly), I've never prioritized fixing it. What are your thoughts? Do you have a better approach in mind?

uptudev commented:
I don't know if you have another mod interfering with your input on your end, but multiselect can be bound perfectly fine as of v2.0.4 on my end beyond a minor visual glitch when attempting to rebind the first time; subsequent attempts to rebind work fine. I also cannot replicate your issue with being unable to trigger the restart feature, as it can be bound to either clicks or holds (unbound by default though, because it's the most potentially destructive bind), and both of them work on my end as well. I even removed my old configuration and tried from a new install and it worked fine. Please let me know if the v2.0.5 changes fixed the issue, as they solved the visual glitches I found.

Also ensure you are holding the key down as indicated when trying to rebind any 'hold-only' binds; 'click'-style features like hand sorting and play/discard can be bound to either keypresses or keyholds, but multiselect must be bound to some keyhold, not a press.

I'll be releasing another minor patch in a few minutes with fixes for the visual glitches in the config menu, but as mentioned earlier I couldn't replicate your issues with multiselect nor with the restart function, so I can't patch it unless you can provide more info such as a crash dump or recording of the erroneous behaviour.

___

Regarding the suggestions:

>     1. Add a "Restore Default Keybinds" button.
>        I've noticed that for many mods, even deleting the mods, mod loader, and config folder, mods's settings still aren't reset. A dedicated reset button would be helpful.

~~I will probably have one in the next minor patch later today, as this is a great QoL addition and it will be trivial to add.~~
Added in [`1aa142e`](https://github.com/Kooluve/Better-Mouse-And-Gamepad/pull/8/commits/1aa142e8a52b05ed959b0ad124802dbc4ea8739c).

> 
>     2. Limit bindable keys to common game keys (letters, numbers, spacebar, etc.).
>        Modifier keys (like Ctrl, Alt) or system keys (Win, CapsLock) can cause unintended issues such as StickyKeys activation or unexpected behavior.

StickyKeys is OS-specific behaviour that can be disabled from within Windows, and should be unless you explicitly require it; there's no need to further limit input options beyond disallowing `Mouse1` and `ESC` to be bound, as putting the line at modifiers seems completely arbitrary compared to just limiting it to only keys that the game can support, as the current implementation does. If someone doesn't want `AltGr` bound to a feature, the solution is to not bind it in the first place, rebind it with another key, or to use the ~~(soon to be implemented)~~ reset button to reset.
 
>     3. Rapid mouse movement during multiselect causes intermittent card selection.
>        This seems to occur because the game updates input on a per-frame basis, and fast movement may be missed. This issue exists in both my version and Handy.
>        Since most users are unlikely to encounter this (or can avoid it by moving more slowly), I've never prioritized fixing it. What are your thoughts? Do you have a better approach in mind?

That is a byproduct of how the Love2D game engine operates; any part of the input handling of the game works through frame updates, and input hooks like this mod are bottlenecked by it. In theory a line could be inscribed from the beginning of your multiselect to the end, and any cards that contact that line could be highlighted, but this would be neither more responsive nor easier to implement (if even possible with the way the game operates). You cannot process input any faster than the computer is already processing the rest of the game without creating a forked process that hooks into the memory of the game, but that would involve creating a project similar to *DFHack* for Balatro; as such, there's no feasible way to really fix it.

uptudev committed:
fix(input): fix inconsistent coding style

modify:
* `./src/libs/input.lua`: fix inconsistent coding style

Signed-off-by: uptu <uptu@uptu.dev>

uptudev committed:
feat(config): add config reset button and fix visual bugs

modify:
* `./localization/en-us.lua`: add localization key for reset button
  text; removed brackets from button text
* `./src/libs/config.lua`: add reset button to config menu; fix buggy
  config functions

Signed-off-by: uptu <uptu@uptu.dev>

uptudev committed:
fix(meta): increment version number

modify:
* `./CHANGELOG.md`: add `v2.0.5` changes
* `./bmag.json`: increment version number to `v2.0.5`

Signed-off-by: uptu <uptu@uptu.dev>

uptudev commented:
> Thank you for your efforts, all the previous issues have been resolved. However, I've noticed a few problems with the keybinds:
> 
> 1. The multiselect keybinding cannot be modified.
> 
> 2. The "restart run" action cannot be triggered with any key.

Let me know if these are still issues for you with [`v2.0.5`](https://github.com/uptudev/bmag/releases/tag/v2.0.5). As far as I can tell all bugs with the menu on my end are gone, and that may have fixed the errors you noticed.

Kooluve commented:
I hadn't realized there was a distinction between clicking and long-pressing in the config interface, as most games allow keybinds to be set with a simple click—regardless of whether the actual in-game action is a click or a hold. Although I noticed the hint text at the top, I didn’t read it carefully and simply followed my intuition when setting the keys. Even though you’ve added a clearer prompt in version 2.0.5, I’m still not entirely sure whether all players will notice this detail.

Because of this, I initially tried to set the multiselect key with a click (rather than a long press), and then attempted to use the reset function (which I’d set with a click) by holding the key down. After understanding the intended behavior, I was able to configure and use the keys correctly. However, I’m still uncertain whether distinguishing between click and long-press during keybinding aligns with most players’ intuition.

Regarding the restriction of bindable keys—aside from the Windows key, which can cause more noticeable issues—other keys don’t really have significant drawbacks. It’s reasonable to let players decide their own key mappings, so the current approach is fine.

As for the rapid movement issue during multiselect, one possible solution is to sample the mouse position at short intervals (e.g., every 10 frames, though this would need testing) and check for collisions with card hitboxes along the path formed by these points. It’s important not to only record the start and end positions, since players sometimes need to select cards in a non-linear pattern (e.g., drawing a wavy line with the mouse). You can explore whether this is feasible to implement, though I agree it may be a relatively low-priority fix.

Other bugs:
1. Clicking "Back" in the config interface still saves the settings automatically. The expected behavior might be to only save changes when explicitly confirming (e.g., with an "Apply" or "Save" button).
2. Keybind Conflict/Glitch: If you assign a key to another function while it's already bound to "Multiselect," so the "Multiselect" keybind is cleared. However, the game continues to perform the multiselect action with that key until you either: a) reset all keybinds, or b) reassign a key to "Multiselect" and use it once.

Enhancement:
In the config interface, the "Suit" option is positioned above the "Value" option. It might be more intuitive to swap their positions (Value above Suit).

uptudev commented:
No.

Quit responding to me using AI prompts. I've been doing software engineering for over a decade and a half, and I don't need someone getting a chatbot to reiterate what I already said, then providing suggestions they cannot even be bothered to verify, *especially* when they regularly hallucinate bugs that don't exist. It's a massive waste of my time as someone trying to contribute this PR for over 2 months.

Has your LLM considered the computational cost of checking a bounding box every few frames may be massive when we've already determined pegging input processing to a potentially slow framerate *is the bottleneck in the hypothetical you're trying to solve*? Parsing input a tenth of the time using collision functions 10x as computationally intensive that probably don't even exist in the engine won't fix anything; Love2D almost certainly doesn't have a way to iterate over every entity making contact with a line, as that would require iterating over every single game object every frame input is parsed. As mentioned earlier when *I* went over the pros and cons, it's neither feasible to implement, nor more responsive, as cards would only be highlighted *every 10 frames*.

Listen to what I'm saying and stop asking an LLM trained on millions of unoptimized web dev portfolios to give you suggestions. I'm not adding more extraneous shit; there are over 90 commits in this pull request. Accept the PR or don't.

Regarding the "bugs", the 'Save' button saves your configuration to disk for any subsequent game launches. It's not an 'Apply' button; any bindings set in the menu are immediately applied, and that is intended behaviour when changing settings in a menu.

I might fix the second bug if it actually exists this time. I have doubts given lines 125-129 and 151-155 in `./src/libs/bindmap.lua` exist explicitly to handle this behaviour, and it's clear you're using AI which clearly hallucinates to review this for you.

> as most games allow keybinds to be set with a simple click—regardless of whether the actual in-game action is a click or a hold

Not when you can bind both clicks and holds to all but one feature; it's called consistency. How many games has the AI that wrote this played again? Oh yeah, zero, so maybe stop asking it.

Kooluve commented:
I apologize for the various delays, careless issue submissions, and unwise suggestions I've caused since you submitted this PR.

I am an amateur developer with only basic computer and programming knowledge, so I did rely heavily on free AI Q&A tools during my mod development process. However, in this PR, all my questions and suggestions were based on my own testing and ideas—at this point I wasn’t being careless. The only part where I used AI was for English translation, as I’m not a native speaker and my English skills are limited. Of course, given my limited knowledge, some of my replies might not even be as good as those from free AI, so I'm sorry about this.

I didn't respond for over a month during this period. My life circumstances weren't great at the time, so I didn't feel like putting effort into this, even though I actually had quite a bit of free time. I truly neglected the results of your hard work, I did it wrong.

Although I only performed simple testing and provided feedback during this PR process, I still learned a lot about GitHub and PRs. Since you are the first person I’ve submitted a PR to on GitHub, I truly appreciate your generosity and enthusiastic guidance. My mod has many imperfect aspects, the code is poorly written, and it's already considered outdated on Discord. I believe your substantial improvements can breathe new life into it.

---

Regarding the rapid movement issue, I admit my consideration was superficial, and I didn’t actually attempt to solve it. As you pointed out, it seems like more trouble than it’s worth. Fortunately, this issue has little practical impact and can be ignored.

---

About the save function: In my experience, most games with an "Apply" button in the keybinding settings require clicking it to save the settings for the current and future sessions. If players don’t click "Apply" and exit directly, the changes aren’t saved. Players might not realize that settings are saved in real time and "Apply" is meant for subsequent game launches. On the other hand, if there’s no "Apply" button, it’s natural to assume settings are saved in real time.

Your current approach combines these two common methods, which feels counterintuitive and requires additional adaptation. If this is a limitation of Love2D or smods, perhaps more eye-catching explanatory text should be added.

---

Regarding the second bug, I can consistently reproduce it (game version, smods, and lovely are all up-to-date, with only bmag installed):

1. Enter the mod configuration menu during a run or from the main menu.
2. Multiselect an existing keybinding (e.g., default settings).
3. Rebind the above hold key to another function.
4. Return to the current run or start a new run.
5. Multiselect will automatically trigger afterward.
6. Three temporary fixes: a) Bind any key to multiselect, return to the run, and hold press that key. b) Click the reset button. c) Restart the game.

If needed, I can provide a video with keypress visualization.

---

About binding keys with clicks and holds: In all the games I’ve played, I’ve never encountered a scenario where holding a key was required for binding. This feels very counterintuitive I think. I also checked several games in my Steam library (like Slay the Spire, Celeste, Monster Hunter or Apex, etc.) and found no such cases. Keybinding typically involves either clicking to input, selecting from options, or assigning a function to a key.

Specifically, in games where clicking and holding perform different actions, the conditions and effects are either:

- Same condition, similar effects (e.g., normal attack vs. charged attack).

- Same conditions, different effects (e.g., tapping Shift to toggle walk/run vs. holding Shift to run and releasing to walk).

- Different conditions, different effects (e.g., key E to interact: tapping to talk to NPCs vs. holding to heal teammates with a medkit).

Thus, even if one key corresponds to multiple functions, only a single key needs to be bound. The distinction between click and hold is handled internally by the game, and users only need to click during keybinding. 

Our mod may belongs to the second category. Combining this with the ability to bind both clicks and holds (which is counterintuitive) further blurs the line between binding input and actual operation.

I think it would be better to restrict multiselect and restart to hold-only actions, while other functions are click-only. This would limit all keybinding inputs to clicks. Given the mod’s limited functions, there’s no need to conserve key slots by using holds. This approach wouldn’t restrict user freedom too much, would be more intuitive, and would avoid unexpected issues.

I'm not sure if I'm overthinking the issue of distinguishing between clicks and holds during keybinding. Although it's easy to adapt once you understand the logic, I did experience significant confusion during my initial use. 

---

I hope my suggestions provide some useful insights. If you’re busy, there’s no need to reply quickly—feel free to respond and optimize at your own pace.☺️

uptudev committed:
fix(smods): fixed steammodded version number causing issues

modify:
* `./bmag.json`: replaced required version with a wildcard, as the
  current Steammodded semver procedures are causing issues recognissing
  beta versions

uptudev committed:
chore(changelog): update to v2.0.6

modify:
* `./CHANGELOG.md`: update changelog with v2.0.6 changes
